### PR TITLE
Add support for partition schema evolution for HUDI tables

### PR DESCRIPTION
**Pull request by imjalpreet**
This is a follow-up PR for #16011. This PR enables partition schema evolution for HUDI tables.

```
== RELEASE NOTES ==

Hive Changes
* Add support for allowing to match columns between table and partition schemas by names for HUDI tables. This is enabled when configuration property ``hive.parquet.use-column-names`` or the hive catalog session property ``parquet_use_column_names`` is set to true. By default they are mapped by index. 
```

### CODE REVIEW

**Review by zhenxiao — COMMENTED** [review]
one minor thing

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 1]
```diff
@@ -385,7 +389,11 @@ public CounterStat getHighMemorySplitSource()
             }
         }
 
-        Optional<HiveStorageFormat> storageFormat = getHiveStorageFormat(table.getStorage().getStorageFormat());
+        StorageFormat storageFormat = table.getStorage().getStorageFormat();
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:392]: shall we check session property or configuration here? according to the release note:
This is enabled when configuration property ``hive.parquet.use-column-names`` or the hive catalog session property ``parquet_use_column_names`` is set to true. By default they are mapped by index.

Comment (imjalpreet) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:392]: @zhenxiao [Line 392](https://github.com/prestodb/presto/pull/16348/files/dc3c96ae0ae86a608bc03d6425b2651cac18e72a#diff-be0abdaf5165467c629fccc676e607333b40ac777c8987a7fdc1c3d52183811dR392) and [Line 393](https://github.com/prestodb/presto/pull/16348/files/dc3c96ae0ae86a608bc03d6425b2651cac18e72a#diff-be0abdaf5165467c629fccc676e607333b40ac777c8987a7fdc1c3d52183811dR393) are required in any case. Only lines [395](https://github.com/prestodb/presto/pull/16348/files/dc3c96ae0ae86a608bc03d6425b2651cac18e72a#diff-be0abdaf5165467c629fccc676e607333b40ac777c8987a7fdc1c3d52183811dR395) and [396](https://github.com/prestodb/presto/pull/16348/files/dc3c96ae0ae86a608bc03d6425b2651cac18e72a#diff-be0abdaf5165467c629fccc676e607333b40ac777c8987a7fdc1c3d52183811dR396) are not mandatory if `hive.parquet.use-column-names` is not set to true. But in any case value of this variable is only used in one method `getTableToPartitionMapping` which has the required check at [line 521](https://github.com/prestodb/presto/pull/16348/files/dc3c96ae0ae86a608bc03d6425b2651cac18e72a#diff-be0abdaf5165467c629fccc676e607333b40ac777c8987a7fdc1c3d52183811dR521). 

Anyways, I have added a check here as well, let me know if you feel it is not necessary.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 2]
```diff
@@ -385,7 +389,11 @@ public CounterStat getHighMemorySplitSource()
...
+        Optional<HiveStorageFormat> hiveStorageFormat = getHiveStorageFormat(storageFormat);
+
+        // Use Hive Storage Format as Parquet if table is of HUDI format
+        Optional<HiveStorageFormat> finalHiveStorageFormat = (!hiveStorageFormat.isPresent() && isHudiFormat(storageFormat)) ? Optional.of(PARQUET) : hiveStorageFormat;
```
[the remark this answers is not in the document]
Comment (imjalpreet) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:396]: Sure, made the requested change.

**Review by imjalpreet — COMMENTED** [review]
@zhenxiao Sorry for the delay, I was away for the past few days. 

I have made the requested changes, please let me know your views.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java (hunk 1 above)
Comment (zhenxiao) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:392]: looks nice

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 3]
```diff
@@ -385,7 +389,18 @@ public CounterStat getHighMemorySplitSource()
             }
         }
 
-        Optional<HiveStorageFormat> storageFormat = getHiveStorageFormat(table.getStorage().getStorageFormat());
+        StorageFormat storageFormat = table.getStorage().getStorageFormat();
+        Optional<HiveStorageFormat> hiveStorageFormat = getHiveStorageFormat(storageFormat);
+
+        Optional<HiveStorageFormat> resolvedHiveStorageFormat;
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:395]: how about:
```
Optional<HiveStorageFormat> resolvedHiveStorageFormat = hiveStorageFormat;
if (isUseParquetColumnNames(session)) {
 ...
}
```

**Review by zhenxiao — APPROVED** [review]
thank you, @imjalpreet 
looks good. once minor issue

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java (hunk 3 above)
Comment (imjalpreet) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:395]: @zhenxiao The variable `resolvedHiveStorageFormat` is being used in a lambda function so it has to be a `final` or an `effectively final` variable. If I set it before the if condition it won't remain an `effectively final` variable as it's value will change for the second time inside the if block. 

Due to this reason I had to use an else block.

Let me know if you think I can improve it some other way.

Comment (zhenxiao) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:395]: I might miss something. which lambda function is used for `resolvedHiveStorageFormat`? seems it is only used in `getTableToPartitionMapping`?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 4]
```diff
@@ -385,7 +389,18 @@ public CounterStat getHighMemorySplitSource()
...
+        if (isUseParquetColumnNames(session)) {
+            // Use Hive Storage Format as Parquet if table is of HUDI format
+            resolvedHiveStorageFormat = (!hiveStorageFormat.isPresent() && isHudiFormat(storageFormat)) ? Optional.of(PARQUET) : hiveStorageFormat;
+        }
+        else {
+            resolvedHiveStorageFormat = hiveStorageFormat;
+        }
 
         Iterable<List<HivePartition>> partitionNameBatches = partitionExponentially(hivePartitions, minPartitionBatchSize, maxPartitionBatchSize);
         Iterable<List<HivePartitionMetadata>> partitionBatches = transform(partitionNameBatches, partitionBatch -> {
```
Comment (imjalpreet) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:391]: @zhenxiao I was talking about this. You are right that resolvedHiveStorageFormat is only being used in the method getTableToPartitionMapping but that method is being called from this lambda function. The call is on the line 466 in the updated code.

**Review by imjalpreet — COMMENTED** [review]
@zhenxiao I have left a comment at the line where the lambda function starts. Can you have a look and let me know in case of any concerns.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java (hunk 4 above)
Comment (zhenxiao) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:391]: get it. you are correct

**Review by zhenxiao — APPROVED** [review]

**Pull request merged**